MudDataGrid: Refresh grouping after InvokeServerLoadFunc, ExpandAllGroups, CollapseAllGroups

MihFig opened the pull request:
## Description
Fixes #9149 

## How Has This Been Tested?
Only manually inside my own project used MudBlazor

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

MihFig committed:
Add GroupItems method after InvokeServerLoadFunc (regrouping after _server_data change)

ScarletKuro commented:
Hi.
Thanks for PR.
Please, add a bUnit test for this case.

MihFig committed:
Add Test for DataGrid grouped with ServerData pagination.
Add GroupItems calls to ExpandAllGroups(), CollapseAllGroups() and make them async

MihFig committed:
Reformat code

ScarletKuro commented:
Thanks for the test.

Now I'm concerned about the `async void` you just added. We don't allow that anymore.

Is the `InvokeAsync` required? If yes, then we should consider breaking the API and make `CollapseAllGroups` / `ExpandAllGroups` return a `Task` and add the `Async` suffix.

MihFig commented:
-Is the InvokeAsync required?
In my opinion yes - GroupItems() calls StateHasChanged() and need to refresh view after click.
-If yes, then we should consider breaking the API and make CollapseAllGroups / ExpandAllGroups return a Task and add the Async suffix.
I am doing it right now.


During test creation I noticed buttons ExpandAll/CollapseAll not working too in this scenario - and they do not working in pasted in issue snippet either

MihFig committed:
ApiBreak - change methods
ExpandAllGroups, CollapseAllGroups => ExpandAllGroupsAsync, CollapseAllGroupsAsync

henon commented:
> > Is the InvokeAsync required?
> 
> In my opinion yes - GroupItems() calls StateHasChanged() and need to refresh view after click.

In your opinion? I'd rather go with hard facts. Did you test without `InvokeAsync` and it didn't work?

If I'm not mistaken `InvokeAsync` is only needed if you are currently not on the render thread, i.e. on a background worker or timer thread or if you are on the render thread but you need to trigger another render for some reason.

ScarletKuro commented:
> If I'm not mistaken `InvokeAsync` is only needed if you are currently not on the render thread.

That's correct. It should be be always on the UI thread when it comes from `OnClick` or any other UI interactivity.
I guess it didn't work in test:
https://github.com/MudBlazor/MudBlazor/blob/782eed7c61ba3ac4b0f723373bea60a7f1e95fb2/src/MudBlazor.UnitTests/Components/DataGridTests.cs#L3604
but in that case the
```
await comp.InvokeAsync(() => comp.Instance.CollapseAllGroupsAsync());
```
should be used.

MihFig committed:
Revert "ApiBreak - change methods"

This reverts commit 782eed7c61ba3ac4b0f723373bea60a7f1e95fb2.

MihFig committed:
Remove InvokeAsync in MudDataGrid.razor.cs (ExpandAllGroups/CollapseAllGroups)

MihFig committed:
Text fixes - add InvokeAsync for dataGrid ExpandAllGroups/CollapseAllGroups

henon reviewed (approved):
LGTM, can be merged
